### PR TITLE
Fixes during GDA mainnet rollout

### PR DESCRIPTION
**Pull request by hellwolf**

### CODE REVIEW

File: packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// solhint-disable not-rely-on-time
+pragma solidity 0.8.19;
+
+import { BeaconProxiable } from "../../upgradability/BeaconProxiable.sol";
+
+/**
+ * @title used on first deployment (upgrade case) of GDA
+ * in order to solve the circular dependency between GDA and SuperfluidPool
+ */
+contract SuperfluidPoolPlaceholder is BeaconProxiable {
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol:11]: You sure we have to use it instead of just using ZERO_ADDRESS?


```
nix-shell$ git grep SuperfluidPoolPlaceholder
packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol:contract SuperfluidPoolPlaceholder is BeaconProxiable {
packages/ethereum-contracts/ops-scripts/deploy-framework.js:        "SuperfluidPoolPlaceholder",
packages/ethereum-contracts/ops-scripts/deploy-framework.js:        SuperfluidPoolPlaceholder,
packages/ethereum-contracts/ops-scripts/deploy-framework.js:            SuperfluidPoolPlaceholder.new,
packages/ethereum-contracts/ops-scripts/deploy-framework.js:            "SuperfluidPoolPlaceholder.new"
```

File: packages/ethereum-contracts/ops-scripts/deploy-deterministically.js [hunk 2]
```diff
@@ -110,6 +110,9 @@ module.exports = eval(`(${S.toString()})()`)(async function (
         data: ContractArtifact.bytecode,
         arguments: deployArgs,
     });
+    if (process.env.GAS_PRICE !== undefined) {
```
Comment (hellwolf) [inline packages/ethereum-contracts/ops-scripts/deploy-deterministically.js:113]: I missed this one, what do you use it for?

File: packages/ethereum-contracts/ops-scripts/libs/common.js [hunk 3]
```diff
@@ -66,6 +67,40 @@ function detectTruffle() {
     return truffleDetected;
 }
 
+// extracts the gas related config for the given network from the truffle config
+// returns an object with the relevant fields set in the config (empty if none)
+//
+// NOTE: This implememtation only works for settings provided by a network specific config,
+// not for settings provided by a wildcard (network_id: "*") config.
+function getGasConfig(networkId) {
```
Comment (hellwolf) [inline packages/ethereum-contracts/ops-scripts/libs/common.js:75]: :hot_face:

File: packages/js-sdk/src/Framework.js [hunk 4]
```diff
@@ -7,6 +7,8 @@ const {batchCall} = require("./batchCall");
 const ConstantFlowAgreementV1Helper = require("./ConstantFlowAgreementV1Helper");
 const InstantDistributionAgreementV1Helper = require("./InstantDistributionAgreementV1Helper");
 const fetch = require("node-fetch");
+// yes, this is dirty. truffle is deprecated, so this won't stick forever.
+const truffleConfig = require("../../ethereum-contracts/truffle-config");
```
Comment (hellwolf) [inline packages/js-sdk/src/Framework.js:11]: :cold_sweat:

File: packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol (hunk 1 above)
Comment (d10r) [reply packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol:11]: we need it in order to have a _compatible logic_:
```
contract SuperfluidUpgradeableBeacon is UpgradeableBeacon {
...
    function upgradeTo(address newImplementation) public override onlyOwner {
...
        if (BeaconProxiable(newImplementation).proxiableUUID() != BeaconProxiable(implementation()).proxiableUUID()) {
            revert INCOMPATIBLE_LOGIC();
        }
```

File: packages/ethereum-contracts/ops-scripts/deploy-deterministically.js (hunk 2 above)
Comment (d10r) [reply packages/ethereum-contracts/ops-scripts/deploy-deterministically.js:113]: to deploy the forwarders at the same address

File: packages/ethereum-contracts/ops-scripts/libs/common.js (hunk 3 above)
Comment (d10r) [reply packages/ethereum-contracts/ops-scripts/libs/common.js:75]: yeah, I hate it too. But deep diving into a deprecated framework in order to figure out something less hacky doesn't seem like a good investment of time.

File: packages/js-sdk/src/Framework.js (hunk 4 above)
Comment (d10r) [reply packages/js-sdk/src/Framework.js:11]: if you can't get over it, I can also keep a local patch with this hack instead of pushing it.

Comment (hellwolf) [reply packages/js-sdk/src/Framework.js:11]: no, it's fine, if you couldn't find a workaround with less copy pasted code, then let's incur another technical debt here.

Comment (d10r) [reply packages/js-sdk/src/Framework.js:11]: I think of it as tech debt we know is about to be written off

File: packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.ts [hunk 5]
```diff
@@ -430,7 +438,9 @@ describe("Superfluid Host Contract", function () {
                     paNFTLogicAddress,
                     pmNFTLogicAddress
                 );
-                await governance.updateContracts(
+                await governance[
+                    "updateContracts(address,address,address[],address,address)"
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.ts]: If you removed the overloadded updateContracts, are these changes still needed.

File: packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol (hunk 1 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol:11]: interesting, checkout the SuperfluidDeployer code, it uses zero address, what's the difference?

File: packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.ts (hunk 5 above)
Comment (d10r) [reply packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.ts]: no, shall we revert those?

Comment (hellwolf) [reply packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.ts]: If you are sure about removing the old method, then it seems unnecessary noise here.

I asked in another thread though,  are you sure? Have you  thought about all the testnets we are still supporting, and any mainnet that might still use the old abi.

File: packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol (hunk 1 above)
Comment (d10r) [reply packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol:11]: there were 2 reasons:
* don't waste gas for deploying a logic which is then immediately replaced

and more importantly:
* have all code paths blocked during the intermediate phase with the new agreement registered, but not yet properly set up (if `SuperfluidPool.initialize()` reverts, then `GeneralDistributionAgreementV1.createPool()` reverts too, thus we know for sure there's no way to use the agreement to manipulate SuperToken state as long as we're not done with the setup).

Comment (d10r) [reply packages/ethereum-contracts/contracts/agreements/gdav1/SuperfluidPoolPlaceholder.sol:11]: we can remove this after upgrading all mainnets, not needed for new deployments

File: packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.ts (hunk 5 above)
Comment (d10r) [reply packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.ts]: yes, I'm sure

**Comment by hellwolf** [comment]
Please let's keep these up to date and merge:

- Changelogs
- packages/ethereum-contracts/audits/2023-12 Trail of Bits - Changes Since Audit.md

**Review by d10r — APPROVED** [review]

**Review by 0xdavinchee — APPROVED** [review]
LGTM
